lfs/sandboxes/heka/input/tail.lua

```diff
@@ -128,84 +129,92 @@ if delimiter then
+        for data in fh:lines_tail(true) do
             local ok, err = pcall(decode, data, default_headers)
             if (not ok or err) and send_decode_failures then
                 err_msg.Payload = err
                 pcall(inject_message, err_msg)
             end
-            checkpoint = fh:seek()
-            inject_message(nil, checkpoint)
+            offset = offset + #data + 1
+            write_checkpoint()
         end
     end
 end
 
 
-local function open_file(checkpoint)
-    local fh, err = io.open(input_filename, "rb")
+local function get_inode()
+    return lfs.attributes(input_filename, "ino")
+end
+
+
+local function open_file()
+    local err
+    fh, err = gzfile.open(input_filename, "rb")
     if not fh then
         print(err)
-        checkpoint = 0
+        return fh
     end
 
-    if checkpoint ~= 0 then
-        local seek = fh:seek("end")
-        if checkpoint > seek or not fh:seek("set", checkpoint) then
-            print(string.format("invalid checkpoint: %d end: %d, starting from the beginning", checkpoint, seek))
-            checkpoint = 0
-        end
+    local cinode = get_inode()
+    if inode ~= cinode then
+        inode = cinode
+        offset = 0
     end
-    inject_message(nil, checkpoint)
-    return fh, checkpoint
-end
 
-
-local function get_inode()
-    return lfs.attributes(input_filename, "ino")
+    if offset ~= 0 then fh:seek("set", offset) end
+    write_checkpoint()
+    return fh
 end
 
 
-local inode
-local function follow_name(fh, checkpoint)
-    if not inode then inode = get_inode() end
-    while true do
-        read_until_eof(fh, checkpoint)
-        local tinode = get_inode()
-        if inode ~= tinode then
-            flush_remaining()
-            inode = tinode
-            fh:close()
-            fh, checkpoint = open_file(0)
-            if not fh then return nil end
-        else
-            return fh -- poll
-        end
+local function follow_name()
+    if not fh then return end
+    read_until_eof()
+    if inode ~= get_inode() then
+        flush_remaining()
+        fh:close()
+        fh = nil
     end
 end
 
 
-local function follow_descriptor(fh, checkpoint)
-    read_until_eof(fh, checkpoint)
-    return fh -- poll
+local function follow_descriptor()
+    if not fh then return end
+    read_until_eof()
 end
 
 
 if follow == "name" then
-    require "lfs"
     follow = follow_name
 else
     follow = follow_descriptor
 end
 
 
-local fh
 function process_message(checkpoint)
-    checkpoint = checkpoint or 0
     if not fh then
-        fh, checkpoint = open_file(checkpoint)
-        if not fh then return 0 end
+        local t = type(checkpoint)
+        if t == "string" then
+            local i, o = t:match("(%d+):(%d+)")
+            inode  = tonumber(i) or 0
+            offset = tonumber(o) or 0
+        elseif t == "number" then  -- migrate the old checkpoint format
+            inode  = get_inode() or 0
+            offset = t
```

Comment:
This probably should be:
```
offset = inode
```
Otherwise `offset` will be assigned string value `"number"`.

Reply:
it needs to be set to checkpoint... fixed